src/main/java/org/zalando/nakadi/service/CursorsService.java

```diff
@@ -91,30 +86,60 @@ public CursorsService(final ZooKeeperHolder zkHolder,
-                                 final String streamId) throws NakadiException {
-        final ZkSubscriptionNode subscription = subscriptionClient.getZkSubscriptionNodeLocked();
-        if (!Arrays.stream(subscription.getSessions()).anyMatch(session -> session.getId().equals(streamId))) {
+    private void validateStreamId(final List<SubscriptionCursor> cursors, final String streamId,
+                                  final String subscriptionId) throws NakadiException {
+
+        if (!isActiveSession(subscriptionId, streamId)) {
             throw new InvalidStreamIdException("Session with stream id " + streamId + " not found");
         }
-        final Map<Partition.PartitionKey, Partition> partitions = Arrays.stream(subscription.getPartitions())
-                .collect(Collectors.toMap(Partition::getKey, Function.identity()));
-        final List<SubscriptionCursor> invalidCursors = cursors.stream()
-                .filter(cursor ->
-                        Try.cons(() -> eventTypeRepository.findByName(cursor.getEventType()))
-                                .getO()
-                                .map(eventType -> {
-                                    final Partition.PartitionKey key =
-                                            new Partition.PartitionKey(eventType.getTopic(), cursor.getPartition());
-                                    final Partition partition = partitions.get(key);
-                                    return partition == null || !streamId.equals(partition.getSession());
-                                })
-                                .orElseGet(() -> false))
-                .collect(Collectors.toList());
-        if (!invalidCursors.isEmpty()) {
-            throw new InvalidStreamIdException("Cursors " + invalidCursors + " cannot be committed with stream id "
-                    + streamId);
+
+        try {
+            final List<SubscriptionCursor> invalidCursors = cursors.stream()
+                    .filter(cursor -> {
+                        try {
+                            final EventType eventType = eventTypeRepository.findByName(cursor.getEventType());
+                            final String partitionSession = getPartitionSession(subscriptionId, eventType.getTopic(),
+                                    cursor.getPartition());
+                            return !streamId.equals(partitionSession);
+                        } catch (final InternalNakadiException | NoSuchEventTypeException |
+                                ServiceUnavailableException e) {
+                            throw new NakadiRuntimeException(e);
+                        }
+                    })
+                    .collect(Collectors.toList());
+
+            if (!invalidCursors.isEmpty()) {
+                throw new InvalidStreamIdException("Cursors " + invalidCursors + " cannot be committed with stream id "
+                        + streamId);
+            }
+        } catch (final NakadiRuntimeException e) {
```

Comment:
I wanted to catch here only what I throw inside lambda. As I understand our general approach is that we don't catch runtime exceptions everywhere (for me it's like to check everything for null).